### DISCUSSION TITLE
Using cx_freeze to create 'portable' version of Python 2.7 program

### DISCUSSION THREAD

**Discussion by chrisisbd** [marcelotduarte/cx_Freeze#845] [Q&A]
I am using cx_freeze to try and create a 'portable' version of a Python 2.7 printer utility, this is purely for my own use.  I use xubuntu on nearly all my systems here and in moving from xubuntu 20.04 to 20.10 and newer versions it's no longer possible to run the Python 2.7 utility in its 'naked' form.  There are a number of reasons for this, it uses GTK which has major changes and it has a .so built for Python 2.7 that I can't recompile for Python 3 (otherwise I can convert the rest of the code to Python 3).

So, I have installed xubuntu 18.04 on a old system I have lying around and I have installed cx_freeze version 5.1.1 (as the last cxfreeze for Python 2.7) on that system.  I have installed my printer utility and it runs OK.

The printer utility is called 'scantool.py' so I have run the following command:-

    cxfreeze /usr/libexec/okimfputl/scantool.py

It runs OK, with some warnings about missing packages (but I think they're all Windows ones) and produces a 'scantool' executable in ~/dist.  That runs OK on the system where I created it but needs some extra files copied across to any target.

Specifically, when I copy the whole of the ~/dist directory across to another system and run ~/.dist/scantool it complains about several missing .so files and data files.  I can fix these problems by copying the files in question across to the target system but is there a way to get the cxfreeze command itself to copy them into the build?  As it is I have to add LD_LIBRARY_PATH to the environment when running scantool and have to put the data files in the same place on the target system as they were on the build system.  It would be handy if I could move them into the ~/dist directory.

So, can I do this with the cxfreeze command?  I.e.:-

     How to specify additional .so files which must be copied into the build?

     How do I specify/change the directories where data files are to be found?  I can copy the files manually but 
     how do I tell cxfreeze where they will be?

Thanks for any help with this.  I'd prefer not to have to create  distutils/setup for this because it is just a one off to get something working on my desktop machine, I'm not wanting to distribute it to any other machine.

**Comment by marcelotduarte** [comment]
I can create a setup.py, the basic is:
```
from cx_Freeze import setup, Executable

executables = [Executable("hello.py")]

setup(
    name="hello",
    version="0.1",
    description="Sample cx_Freeze script",
    executables=executables,
)
```
Please see the docs, you can need the build_exe option like includes, packages and/or include_files.
Then freeze by using python setup.py build

**Comment by chrisisbd** [comment]
I'm completely lost!  All this stuff about setup.py and build_exe and so on makes no sense to me at all.

I'm by no means completely naive about computers but there seems to be a missing link somewhere here that I can't understand.

Given that I have a python program "scantool.py" which I want to make transportable as a package to systems that (possibly) don't have python installed what do I do?

Running "cxfreeze /usr/libexec/okimfputl/scantool.py"  actually does a pretty good job and I can add various missing bits to get a working result.

However all the documentation and help here talks about setup.py and build_exe and I'm completely lost!  I'm sure there's something very simple I'm missing somewhere but at the moment I can't see it.

What I don't want or need is a python package to install somewhere, I just want to transport a single python program to a system which doesn't have any of the normal python support for its version of python.  I get the feeling that all the stuff about setup.py and build_exe is to produce something that can be installed with pip or similar, that's not what I'm after.

**Reply by chrisisbd** [reply]
Oops, maybe saying ".... make transportable as a package to...." is the wrong wording.  I don't want an installable package, I just want something I can *run* on the target system.

**Comment by marcelotduarte** [comment]
Running cxfreeze is the basic tool. cxfreeze-quickstart is another tool that creates a basic setup.py.
Please read the documentation.

**Comment by marcelotduarte** [comment]
You can run cxfreeze --help to a basic help.
